### PR TITLE
C++: make internal Objective-C API wrap `DocumentSnapshot` from core API

### PR DESCRIPTION
**Pull request by var-const**
Notes:
* this also adds support of Objective-C classes to `util::Hash`;
* I think we need a good distinctive name for the new layer. Calling it "api" is a big source of confusion, because the Objective-C `API/` already exists and delegates to it. The best I could come up with so far is "shim", which according to [Wikipedia](https://en.wikipedia.org/wiki/Shim_(computing)) means _"a library that transparently intercepts API calls and changes the arguments passed, handles the operation itself or redirects the operation elsewhere"_, which seems similar to what this layer does, translating two public APIs to an internal representation. I'm very open to suggestions.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/util/hashing.h [hunk 1]
```diff
@@ -84,7 +87,7 @@ struct has_std_hash {
  */
 template <typename T>
 using std_hash_type =
-    typename std::enable_if<has_std_hash<T>::value, size_t>::type;
+    typename absl::enable_if_t<has_std_hash<T>::value, size_t>;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/util/hashing.h]: This is just slightly nicer because it saves having to type `::value`.

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 2]
```diff
@@ -88,79 +90,55 @@ + (instancetype)snapshotWithFirestore:(FIRFirestore *)firestore
...
                          document:(nullable FSTDocument *)document
                         fromCache:(BOOL)fromCache
                  hasPendingWrites:(BOOL)pendingWrites {
   if (self = [super init]) {
-    _firestore = firestore;
-    _internalKey = std::move(documentKey);
-    _internalDocument = document;
-    _fromCache = fromCache;
-    _pendingWrites = pendingWrites;
+    _shim = DocumentSnapshot{firestore, std::move(documentKey), document, fromCache, pendingWrites};
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: As discussed via chat, "shim" isn't quite the right concept here. A shim is something that might sit in between the Objective-C API and the C++ core API, adjusting parameters to fit. If anything the Objective-C wrapper is more of a shim than the the underlying DocumentSnapshot which is the actual implementation.

Ignoring Cocoa's use of the term, this is more like a "delegate". See for example, the way that term is used in e.g. [`ForwardingList`](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ForwardingList.java#L61).

Unfortunately, in Cocoa, a "delegate" has a [narrower meaning](https://developer.apple.com/library/archive/documentation/General/Conceptual/DevPedia-CocoaCore/Delegation.html).

In Guava's usage a delegate is a sort of underlying object that's supposed to do all the work. In Cocoa's usage, a delegate is a cooperating object that gets to receive notifications and specify policy without being a part of the implementation. While the former fits, the latter doesn't 😖.

In this particular case it doesn't happen, but in many others it seems as if we'll have multiple members that need to use (e.g. the firestore reference and something else). In those cases calling one the delegate "delegate" or "shim" and not the other one will be unclear.

All of which is just to say, how about we just avoid the general naming problem of what to call this kind of underlying object and just name this `_snapshot`?

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 3]
```diff
@@ -191,27 +171,24 @@ - (nullable id)valueForField:(id)field
...
+  FSTFieldValue *fieldValue = _shim.GetValue(fieldPath.internalValue);
   return fieldValue == nil ? nil : [self convertedValue:fieldValue options:options];
 }
 
-- (FSTFieldValueOptions *)optionsForServerTimestampBehavior:
+- (nullable id)objectForKeyedSubscript:(id)key {
+  return [self valueForField:key];
+}
+
+- (FSTFieldValueOptions *)createOptionsForServerTimestampBehavior:
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: `optionsForServerTimestampBehavior` was the better name for this according to the [style guide](http://google.github.io/styleguide/objcguide.html#objective-c-method-names) (see the `sandwich` example). Please revert.

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 4]
```diff
@@ -220,18 +197,20 @@ - (id)convertedValue:(FSTFieldValue *)value options:(FSTFieldValueOptions *)opti
   } else if ([value isKindOfClass:[FSTReferenceValue class]]) {
     FSTReferenceValue *ref = (FSTReferenceValue *)value;
     const DatabaseId *refDatabase = ref.databaseID;
-    const DatabaseId *database = self.firestore.databaseID;
+    const DatabaseId *database = _shim.firestore().databaseID;
     if (*refDatabase != *database) {
       // TODO(b/32073923): Log this as a proper warning.
-      NSLog(@"WARNING: Document %@ contains a document reference within a different database "
-             "(%s/%s) which is not supported. It will be treated as a reference within the "
+      NSLog(@"WARNING: Document %@ contains a document reference within a "
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: These lines seem like they shouldn't have changed.

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 5]
```diff
@@ -88,79 +90,55 @@ + (instancetype)snapshotWithFirestore:(FIRFirestore *)firestore
...
 
 @implementation FIRDocumentSnapshot {
-  FIRSnapshotMetadata *_cachedMetadata;
-  DocumentKey _internalKey;
+  DocumentSnapshot _shim;
 }
 
-@dynamic metadata;
-
 - (instancetype)initWithFirestore:(FIRFirestore *)firestore
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: Since the guts of this type are now contained in the `DocumentSnapshot`, could we make this initializer be `initWithSnapshot:(DocumentSnapshot &&)snapshot` instead?

That would make the factory method `snapshotWithSnapshot:(DocumentSnapshot)snapshot` too and drastically cut down on the verbosity of all this parameter passing.

Note:
  * We use this factory method technique here to hide the initializer from the public interface. You can add arbitrary methods but you can't add designated initializers.
  * The initializer can take a move-only parameter because the factory shouldn't ever do anything but move.

File: Firestore/core/src/firebase/firestore/api/document_snapshot.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      : firestore_{firestore},
+        internal_key_{std::move(document_key)},
+        internal_document_{document},
+        from_cache_{from_cache},
+        has_pending_writes_{has_pending_writes} {
+  }
+
+  size_t Hash() const;
+
+  bool Exists() const;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/api/document_snapshot.h]: Seems as if `Exists` and `GetInternalDocument` could just be accessors, specified inline?

File: Firestore/core/src/firebase/firestore/api/document_snapshot.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  FSTDocument* GetInternalDocument() const;
+  FIRDocumentReference* CreateReference() const;
+  std::string GetDocumentId() const;
+  FIRSnapshotMetadata* GetMetadata() const;
+
+  FSTObjectValue* GetData() const;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/api/document_snapshot.h]: For nullable Objective-C types, should we preserve nullability annotations?

`nullable FSTObjectValue*` is unlikely to work, but `FSTObjectValue _Nullable*` probably would.

File: Firestore/core/src/firebase/firestore/api/document_snapshot.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+@class FIRFirestore;
+@class FIRSnapshotMetadata;
+@class FIRDocumentReference;
+@class FSTDocument;
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+class DocumentSnapshot {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/api/document_snapshot.h]: High-level comment: you seem to be trying to hide the implementation of this class, but I think it's unnecessary. This isn't a public interface.

In any case where the body of a method can be simply specified inline, wouldn't it be easier to just do so?

File: Firestore/Source/API/FIRDocumentSnapshot.mm (hunk 5 above)
Comment (var-const) [reply Firestore/Source/API/FIRDocumentSnapshot.mm]: Done. Note that I also implemented `QueryDocumentSnapshot` now, so that it now takes a (unique) pointer to `DocumentSnapshot`.

File: Firestore/Source/API/FIRDocumentSnapshot.mm (hunk 4 above)
Comment (var-const) [reply Firestore/Source/API/FIRDocumentSnapshot.mm]: Done.

File: Firestore/Source/API/FIRDocumentChange.mm [hunk 9]
```diff
@@ -63,12 +66,11 @@ + (FIRDocumentChangeType)documentChangeTypeForChange:(const DocumentViewChange &
...
     for (const DocumentViewChange &change : snapshot.documentChanges) {
-      FIRQueryDocumentSnapshot *document = [FIRQueryDocumentSnapshot
-          snapshotWithFirestore:firestore
-                    documentKey:change.document().key
-                       document:change.document()
-                      fromCache:snapshot.isFromCache
-               hasPendingWrites:snapshot.mutatedKeys.contains(change.document().key)];
+      auto apiDocument = absl::make_unique<QueryDocumentSnapshot>(
+          firestore, change.document().key, change.document(), snapshot.isFromCache,
+          snapshot.mutatedKeys.contains(change.document().key));
```
Comment (rsgowman) [inline Firestore/Source/API/FIRDocumentChange.mm]: You might consider a named parameter for this last one too.
Or you might create a temporary for it: `bool hasPendingWrites = ^;` It's enough of a mouthful to warrant it (IMO)

Both optional.

File: Firestore/Source/API/FIRDocumentChange.mm [hunk 10]
```diff
@@ -90,12 +92,11 @@ + (FIRDocumentChangeType)documentChangeTypeForChange:(const DocumentViewChange &
...
 
-      FIRQueryDocumentSnapshot *document = [FIRQueryDocumentSnapshot
-          snapshotWithFirestore:firestore
-                    documentKey:change.document().key
-                       document:change.document()
-                      fromCache:snapshot.isFromCache
-               hasPendingWrites:snapshot.mutatedKeys.contains(change.document().key)];
+      auto apiDocument = absl::make_unique<QueryDocumentSnapshot>(
+          firestore, change.document().key, change.document(), snapshot.isFromCache,
+          snapshot.mutatedKeys.contains(change.document().key));
```
Comment (rsgowman) [inline Firestore/Source/API/FIRDocumentChange.mm]: (as above)

File: Firestore/Source/API/FIRQuerySnapshot.mm [hunk 11]
```diff
@@ -129,13 +137,10 @@ - (NSInteger)count {
...
-      [result
-          addObject:[FIRQueryDocumentSnapshot
-                        snapshotWithFirestore:firestore
-                                  documentKey:document.key
-                                     document:document
-                                    fromCache:fromCache
-                             hasPendingWrites:self.snapshot.mutatedKeys.contains(document.key)]];
+      auto apiResult =
+          absl::make_unique<QueryDocumentSnapshot>(firestore, document.key, document, fromCache,
+                                              self.snapshot.mutatedKeys.contains(document.key));
```
Comment (rsgowman) [inline Firestore/Source/API/FIRQuerySnapshot.mm]: (as above)

File: Firestore/Source/API/FIRTransaction.mm [hunk 12]
```diff
@@ -127,19 +130,16 @@ - (void)getDocument:(FIRDocumentReference *)document
         HARD_ASSERT(documents.size() == 1, "Mismatch in docs returned from document lookup.");
         FSTMaybeDocument *internalDoc = documents.front();
         if ([internalDoc isKindOfClass:[FSTDeletedDocument class]]) {
-          FIRDocumentSnapshot *doc = [FIRDocumentSnapshot snapshotWithFirestore:self.firestore
-                                                                    documentKey:document.key
-                                                                       document:nil
-                                                                      fromCache:NO
-                                                               hasPendingWrites:NO];
+          auto apiDoc =
+              absl::make_unique<DocumentSnapshot>(self.firestore, document.key, nil, false, false);
```
Comment (rsgowman) [inline Firestore/Source/API/FIRTransaction.mm]: As above, but note that this isn't test code, so the reader's context might not be the same. (In other words; still optional, but perhaps a bit less so.)

Also the nil (in addition to the bools)

There's a few more instances below too (both in this file and others).

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 13]
```diff
@@ -261,41 +223,15 @@ - (id)convertedValue:(FSTFieldValue *)value options:(FSTFieldValueOptions *)opti
 
 @interface FIRQueryDocumentSnapshot ()
 
-- (instancetype)initWithFirestore:(FIRFirestore *)firestore
-                      documentKey:(DocumentKey)documentKey
-                         document:(FSTDocument *)document
-                        fromCache:(BOOL)fromCache
-                 hasPendingWrites:(BOOL)pendingWrites NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithSnapshot:(std::unique_ptr<QueryDocumentSnapshot>)snapshot
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: This is borderline broken. I can see how it happens to work here but there are contexts in which it won't so I'm wary.

The issue is that Objective-C methods don't support overloading by type. There isn't any name mangling mechanism so there's no run-time distinction between this method and the superclass method you're overriding. This means that generally, overriding a method with a different signature can lead to some very bad behavior, especially when passing by value.

Now, couple that with the fact that the only caller of this is `+[FIRDocumentSnapshot snapshotWithSnapshot]` and you'll find out that this is not type sound: that caller accepts `std::unique_ptr<DocumentSnapshot>` and happens to be dynamically invoking this method because `[[self class] alloc]` happened to return an instance of `FIRQueryDocumentSnapshot`, but there's nothing guaranteeing the instance it passes is of this narrower type.

So, this appears to guarantee type safety, but it's really lying because you could call `+[FIRQueryDocumentSnapshot snapshotWithSnapshot:std::move(snap)]` with just a plain `std::unique_ptr<api::DocumentSnapshot>` and it would compile and run just fine. This happens to work because both types have an identical representation on the stack and we're not doing anything dependent on the type of `QueryDocumentSnapshot`, but this isn't a feature we should count on.

Separately, I think we're technically in violation of the declaration of `FIRQueryDocumentSnapshot` (and if you look closely at the build or declaration of in `FIRDocumentSnapshot.h` in Xcode you may see a warning). That declaration says that `FIRQueryDocumentSnapshot` provides its own definitions of `-data` and `-dataWithServerTimestampBehavior:` but we're not actually providing them here. This happens to work at runtime though because the instance _does_ actually have those methods, but the base class could conceivably change and this would break.

To resolve these issues:
  * Remove this initializer.
  * (maybe) add `+ (instancetype)snapshotWithSnapshot:(std::unique_ptr<QueryDocumentSnapshot>)snapshot;`. This would make calling the factory method actually type sound. However, see next points.
  * (maybe) add back the methods with different nullability here
  * (maybe) just eliminate `api::QueryDocumentSnapshot` altogether

This last point is perhaps the most interesting because it allows us to avoid making the underlying snapshot a unique pointer again and leaves the nullability fudging we're doing here as part of the language binding rather than as a part of the core API. Note that C++ does not encode nullability, so it's not clear what, if any, benefit there would be from having a public `QueryDocumentDocument` type anyway.

Comment (var-const) [reply Firestore/Source/API/FIRDocumentSnapshot.mm]: > just eliminate `api::QueryDocumentSnapshot` altogether

I think this is the best approach, that type seems more trouble than it's worth.

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 14]
```diff
@@ -28,139 +30,102 @@
...
-
-- (BOOL)isEqualToSnapshot:(nullable FIRDocumentSnapshot *)snapshot {
-  if (self == snapshot) return YES;
-  if (snapshot == nil) return NO;
-
-  return [self.firestore isEqual:snapshot.firestore] && self.internalKey == snapshot.internalKey &&
-         (self.internalDocument == snapshot.internalDocument ||
-          [self.internalDocument isEqual:snapshot.internalDocument]) &&
-         self.pendingWrites == snapshot.pendingWrites && self.fromCache == snapshot.fromCache;
+  const auto &rhs = static_cast<FIRDocumentSnapshot *>(other)->_snapshot;
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/Source/API/FIRDocumentSnapshot.mm]: No longer applicable.

File: Firestore/Source/API/FIRQuerySnapshot.mm [hunk 15]
```diff
@@ -129,13 +137,10 @@ - (NSInteger)count {
...
     NSMutableArray<FIRQueryDocumentSnapshot *> *result = [NSMutableArray array];
     for (FSTDocument *document in documentSet.documentEnumerator) {
-      [result
-          addObject:[FIRQueryDocumentSnapshot
-                        snapshotWithFirestore:firestore
-                                  documentKey:document.key
-                                     document:document
-                                    fromCache:fromCache
-                             hasPendingWrites:self.snapshot.mutatedKeys.contains(document.key)]];
+      auto apiResult = absl::make_unique<QueryDocumentSnapshot>(
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/Source/API/FIRQuerySnapshot.mm]: Reverted this change, so that calling code stays the same. I tried doing a test helper, but I found that there already exists a `FSTTestDocSnapshot` helper that takes different parameters, and distinguishing between the two seemed too hard to be worth it.

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 16]
```diff
@@ -18,6 +18,7 @@
 
 #include <utility>
 
+#import "FIRFirestore.h"
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: This is in the wrong spot. Group with `FIRFirestoreSettings.h` below.

File: Firestore/Source/API/FIRDocumentSnapshot.mm [hunk 17]
```diff
@@ -299,6 +250,6 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
   return data;
 }
 
-@end
+end
```
Comment (wilhuff) [inline Firestore/Source/API/FIRDocumentSnapshot.mm]: Revert this change? (It should be `@end`).

File: Firestore/core/src/firebase/firestore/api/document_snapshot.h [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+      : firestore_{firestore},
+        internal_key_{std::move(document_key)},
+        internal_document_{document},
+        from_cache_{from_cache},
+        has_pending_writes_{has_pending_writes} {
+  }
+
+  size_t Hash() const;
+
+  bool Exists() const {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/api/document_snapshot.h]: Wouldn't the style guide call for these to be `exists()`, `internal_document()`, and `document_id()`?

Comment (var-const) [reply Firestore/core/src/firebase/firestore/api/document_snapshot.h]: Done.

I think it's optional in the style guide. I personally don't really like it, because it looks inconsistent, and also because it seems to create a contract of sorts (presumably if the member function stops being a simple getter, it would have to be renamed).

File: Firestore/Source/API/FIRDocumentSnapshot.mm (hunk 17 above)
Comment (var-const) [reply Firestore/Source/API/FIRDocumentSnapshot.mm]: Done.

File: Firestore/Source/API/FIRDocumentSnapshot.mm (hunk 16 above)
Comment (var-const) [reply Firestore/Source/API/FIRDocumentSnapshot.mm]: Done.
